### PR TITLE
Features/ic

### PR DESCRIPTION
**Pull request by alejandrogonzalvo**
### Input Capture Service
The IC service abstracts configuration of the IC mode of the timer peripheral on the stm32H723.

_**Testing**_
It has been proven to work on duty cycles ranging from 1% to 99%.
IC2

It has been proven to work on frequencies up to 4800Hz.
IC2

It has been detected an error due to a whole cycle being skipped.
Real frequency 10000Hz. Freq Shown 5000Hz.
ICerror

All the testing has been done connecting a PWM to an IC configured on different timers.

_**Usage**_
In order to use it you register the IC instance by passing a pin to the IC::register_ic(Pin) function. 
Then you call Pin::Start after registering all your peripherals.
Finally, you just turn it on. See the example used for testing:
image

### CODE REVIEW

**Comment by StefanCostea** [comment]
Post a pic of your ioc configuration please

**Comment by alejandrogonzalvo** [comment]
PWM:
image
image
image

IC:
image
image
image
image

File: Src/HALAL/Services/InputCapture/InputCapture.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		{HAL_TIM_ACTIVE_CHANNEL_3, TIM_CHANNEL_3},
+		{HAL_TIM_ACTIVE_CHANNEL_4, TIM_CHANNEL_4},
+		{HAL_TIM_ACTIVE_CHANNEL_5, TIM_CHANNEL_5},
+		{HAL_TIM_ACTIVE_CHANNEL_6, TIM_CHANNEL_6}
+};
+
+map<Pin, TimerChannelRisingFalling> IC::pin_timer_map = {
+		{PA0, {&htim2, TIM_CHANNEL_1, TIM_CHANNEL_2}}
+};
+map<uint8_t,Pin> IC::service_ids = {};
```
Comment (PabloGonzalezSegarra) [inline Src/HALAL/Services/InputCapture/InputCapture.cpp:25]: Te falta un espacio
```suggestion
map<uint8_t, Pin> IC::service_ids = {};
```

File: Src/HALAL/Services/InputCapture/InputCapture.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	Pin pin = IC::service_ids[id];
+	TimerChannelRisingFalling tim_ch = IC::pin_timer_map[pin];
+	IC::data ic_data = IC::data_map[tim_ch];
+	return ic_data.duty_cycle;
+}
+
+ChannelsRisingFalling IC::find_other_channel(uint32_t channel) {
+	for (auto pin_timer : IC::pin_timer_map) {
+		if(pin_timer.second.channel_rising == channel || pin_timer.second.channel_falling == channel) {
+			return {pin_timer.second.channel_rising, pin_timer.second.channel_falling};
```
Comment (PabloGonzalezSegarra) [inline Src/HALAL/Services/InputCapture/InputCapture.cpp:84]: Esto esta feo molaría un refactor de alguna manera

File: Src/HALAL/Services/InputCapture/InputCapture.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		return;
+	}
+
+	switch (ic_data.count) {
+		case 0: // First Rising edge
+			ic_data.counter_values[0] = HAL_TIM_ReadCapturedValue(tim_ch.timer, tim_ch.channel_rising);
+			ic_data.count++;
+
+
+			break;
```
Comment (PabloGonzalezSegarra) [inline Src/HALAL/Services/InputCapture/InputCapture.cpp:113]: te sobran lineas en blanco

File: Src/HALAL/Services/InputCapture/InputCapture.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+			ic_data.frequency = refClock/diff;
+
+			float duty_diff;
+			if (ic_data.counter_values[1] > ic_data.counter_values[0]) {
+				duty_diff = ic_data.counter_values[1] - ic_data.counter_values[0];
+
+			}
+
+			// Overflow handling
+			else if (ic_data.counter_values[1] > ic_data.counter_values[2]) {
```
Comment (PabloGonzalezSegarra) [inline Src/HALAL/Services/InputCapture/InputCapture.cpp:149]: no dejes lieneas en blanco que no se entiende nada

**Review by PabloGonzalezSegarra — COMMENTED** [review]
Revisa mis comentarios

**Review by StefanCostea — CHANGES_REQUESTED** [review]

File: Src/HALAL/Services/InputCapture/InputCapture.cpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+			uint32_t refClock = HAL_RCC_GetPCLK1Freq()*2 / (tim_ch.timer->Init.Prescaler+1);
+
+			uint32_t diff;
+
+			if (ic_data.counter_values[2] > ic_data.counter_values[0]) {
+				diff = ic_data.counter_values[2] - ic_data.counter_values[0];
+			}
+
+			// Overflow handling
+			else if (ic_data.counter_values[0] > ic_data.counter_values[2]) {
```
Comment (PabloGonzalezSegarra) [inline Src/HALAL/Services/InputCapture/InputCapture.cpp:139]: no dejes else if con líneas en blanco encima, entiendo que quieras fardar de que tu código tienes muchas líneas pero las que esta en blanco no cuentan

Comment (alejandrogonzalvo) [reply Src/HALAL/Services/InputCapture/InputCapture.cpp:139]: bruh

**Comment by StefanCostea** [comment]
Implementalo como en el PR que propusiste, echale un ojo a https://controllerstech.com/pwm-input-in-stm32/  aqui la implementacion que hace es mucho mas simple

**Pull request closed without merging**